Add various dev tools to CryptoTigers and migrate to Typescript

Melvillian opened the pull request:
This commit adds all of the tools given in
https://github.com/PaulRBerg/solidity-template 's template,
in order for prework students start out with the best developer
environment Hardhat has to offer.

Some questionable changes I would like feedback on are:

- converting to Typescript, some folks think it's too complicate :(
- requiring the user install a version of node >=0.14.12 (for the git
  hooks)
- changing the solidity version to 0.8.9 (more up-to-date)

Melvillian committed:
Add various dev tools to CryptoTigers and migrate to Typescript

This commit adds all of the tools given in
https://github.com/PaulRBerg/solidity-template 's template,
in order for prework students start out with the best developer
environment Hardhat has to offer.

Some questionable changes I would like feedback on are:

- converting to Typescript, some folks think it's too complicate :(
- requiring the user install a version of node >=0.14.12 (for the git
  hooks)
- changing the solidity version to 0.8.9 (more up-to-date)

alex-sumner commented:
Looking at the changes in the contract and test source code, there's nothing I see as problematic. Updating the Solidity version is fine. Using uint256 rather than uint is fine, so is putting semi-colons at the end of every line.

I'm not convinced of the benefits of adding all the tooling at this point. In my day to day work I use yarn rather than npm, and typescript (with semi-colons!) rather than javascript, and typechain etc. But here they add a lot of extra stuff to install and understand without adding much value. The actual contract and test code are virtually unchanged. I would favour keeping things as simple as possible at this point, so prospective students can concentrate on the task in hand. Setting up a fully featured dev environment can come later. By putting it here we force the candidate to understand 4 or 5 things at once, rather than just the 2 or 3 we really need.

Incidentally, the following comment has found its way into the readme prior to this commit: `Hardhat is essential for web3 development`. I think this would meet with strong disagreement from a number of very experienced web3 developers, specifically from the developers and proponents of Brownie. It's just not true, you can develop and test web3 code perfectly well without ever writing a line of Typescript/Javascript, let alone using Hardhat. We have chosen a particular toolset. It's a perfectly good choice, and it's the probably the most  popular choice, but we shouldn't claim it's the only choice.

gabrieljim commented:
Yup I agree with you, I didn't phrase it properly, maybe that could be changed to `Using a proper development environment like Hardhat is essential for web3 development`?

alex-sumner commented:
> That sounds good to me!

Melvillian committed:
Response to alex-sumner and gabrieljim comments

removed all dev tools except for Typescript and typechain.
Also reword the necessity of Hardhat
